classes/context.php

```diff
@@ -0,0 +1,218 @@
+        manager::FLAME_ON_PARAM_NAME,
+        manager::FLAME_OFF_PARAM_NAME,
+    ];
+
+    const REDACTLIST = [
+        'sesskey',
+    ];
+
+    /**
+     * Names of scripts that require more info for grouping.
+     * TODO: This list is not yet complete.
+     */
+    const SCRIPT_NAMES_FOR_GROUPBY = [
+        'admin/settings.php',
+        'admin/search.php',
+        'admin/category.php',
+        'lib/javascript.php',
+    ];
+
+    /**
+     * Gets the script type of the request.
+     *
+     * @return int
+     */
+    public static function get_script_type(): int {
+        if (defined('CLI_SCRIPT') && CLI_SCRIPT) {
+            return profile::SCRIPTTYPE_CLI;
+        } else if (defined('AJAX_SCRIPT') && AJAX_SCRIPT) {
+            return profile::SCRIPTTYPE_AJAX;
+        } else if (defined('WS_SERVER') && WS_SERVER) {
+            return profile::SCRIPTTYPE_WS;
+        }
+        return profile::SCRIPTTYPE_WEB;
+    }
+
+    /**
+     * Gets the parameters given to the request.
+     *
+     * @param int $type - The type of call (cli, web, etc)
+     * @return string For non-cli requests, the parameters are returned in a url query string.
+     *               For cli requests, the arguments are returned in a space sseparated list.
+     */
+    public static function get_parameters(int $type): string {
+        if ($type == profile::SCRIPTTYPE_CLI) {
+            return implode(' ', array_slice($_SERVER['argv'], 1));
+        } else {
+            $parameters = [];
+            parse_str($_SERVER['QUERY_STRING'], $parameters);
+            return http_build_query(self::stripparameters($parameters), '', '&');
+        }
+    }
+
+    /**
+     * Removes any parameter on profile::DENYLIST.
+     *
+     * @param array $parameters
+     * @return array
+     */
+    public static function stripparameters(array $parameters): array {
+        $parameters = array_filter(
+            $parameters,
+            function($i) {
+                return !in_array($i, self::DENYLIST);
+            },
+            ARRAY_FILTER_USE_KEY
+        );
+
+        foreach ($parameters as $i => &$v) {
+            if (in_array($i, self::REDACTLIST)) {
+                $v = '';
+            }
+        }
+
+        return $parameters;
+    }
+
+    /**
+     * Gets the name of the script.
+     *
+     * @return string the request path for this profile.
+     */
+    public static function get_request(): string {
+        global $SCRIPT;
+        // If set, it will trim off the leading '/' to normalise web & cli requests.
+        $request = isset($SCRIPT) ? ltrim($SCRIPT, '/') : profile::REQUEST_UNKNOWN;
+        return $request;
+    }
+
+    /**
+     * Gets a value to be used for grouping profiles.
+     *
+     * @param string $request
+     * @param string $pathinfo
+     * @param string $params
+     * @return string
+     */
+    public static function get_groupby_value(string $request, string $pathinfo, string $params): string {
```

Comment:
This is the meat of the whole patch, it would be good to see see unit tests around it with a dataProvider

There logic here is going to end up having a bunch of logic depends on the scripts. In most scripts we will only use the script name, but in some we will use the script + the params, but we will redact the values of the params. In other cases we will redact values in the query path too.

This isn't quite what we want but its what I've used when do some log analysis which is vague along the same lines.

https://github.com/brendanheywood/scripts/blob/master/log-redact

Mostly we just need a bunch of examples to put into the dataProvider to flesh it out, one example:

/pluginfile.php/1/core_admin/logocompact/300x300/1641856751/moodle-icon.jpg

should be grouped as:

/pluginfile.php/x/core_admin/logocompact/xxx

/pluginfile.php/7282414/mod_book/chapter/644676/electrophoresis%202.PNG

as

pluginfile.php/x/mod_book/chapter/xxx